lute/crypto/src/crypto.cpp

```diff
@@ -89,13 +91,89 @@ int lua_digest(lua_State* L)
     const env_md_st* hashFunction = getHashFunction(L, 1);
     BinaryData message = extractData(L, 2);
 
-    void* buffer = lua_newbuffer(L, EVP_MD_size(hashFunction));
-    if (EVP_Digest(message.data, message.length, (uint8_t*)buffer, nullptr, hashFunction, nullptr) == 0)
+    uint8_t* buffer = static_cast<uint8_t*>(lua_newbuffer(L, EVP_MD_size(hashFunction)));
+    if (EVP_Digest(message.data, message.length, buffer, nullptr, hashFunction, nullptr) == 0)
         luaL_error(L, "%s: failed to compute hash", kDigestName);
 
     return 1;
 }
 
+// seal(message: string | buffer): { ciphertext: buffer, key: buffer, nonce: buffer }
+int lua_secretbox_seal(lua_State* L)
+{
+    int argumentCount = lua_gettop(L);
+    if (argumentCount != 1)
+        luaL_error(L, "%s: expected 1 argument, but got %d", kSealName, argumentCount);
+
+    BinaryData message = extractData(L, 1);
+
+    lua_createtable(L, 0, 3);
+
+    // ciphertext
+    uint8_t* buffer = static_cast<uint8_t*>(lua_newbuffer(L, crypto_secretbox_macbytes() + message.length));
+    lua_setfield(L, -2, kCiphertextField);
+
+    // key
+    uint8_t* key = static_cast<uint8_t*>(lua_newbuffer(L, crypto_secretbox_keybytes()));
+    crypto_secretbox_keygen(key);
+    lua_setfield(L, -2, kKeyField);
+
+    // nonce
+    uint8_t* nonce = static_cast<uint8_t*>(lua_newbuffer(L, crypto_secretbox_noncebytes()));
+    randombytes_buf(nonce, crypto_secretbox_noncebytes());
+    lua_setfield(L, -2, kNonceField);
+
+    // compute the ciphertext based on the message, nonce, and key
+    crypto_secretbox_easy(buffer, static_cast<const uint8_t*>(message.data), message.length, nonce, key);
+
+    // freeze the result
+    lua_setreadonly(L, -1, true);
+
+    return 1;
+}
+
+// open(secretbox: { ciphertext: buffer, key: buffer, nonce: buffer }): buffer
+int lua_secretbox_open(lua_State* L)
+{
+    int argumentCount = lua_gettop(L);
+    if (argumentCount != 1)
+        luaL_error(L, "%s: expected 1 argument, but got %d", kOpenName, argumentCount);
+
+    // read all three of the required fields out of the table
+    lua_getfield(L, 1, "ciphertext");
+    lua_getfield(L, 1, "nonce");
+    lua_getfield(L, 1, "key");
+
+    BinaryData ciphertext = extractData(L, 2);
+    size_t nonceLength = crypto_secretbox_keybytes();
+    uint8_t* nonce = static_cast<uint8_t*>(luaL_checkbuffer(L, 3, &nonceLength));
+    size_t keyLength = crypto_secretbox_keybytes();
+    uint8_t* key = static_cast<uint8_t*>(luaL_checkbuffer(L, 4, &keyLength));
```

Comment:
This will not be right.
The third parameter of `luaL_checkbuffer` is an 'out' parameter (like in `luaL_checklstring`).

```suggestion
    size_t nonceLength = 0;
    uint8_t* nonce = static_cast<uint8_t*>(luaL_checkbuffer(L, 3, &nonceLength));
    if (nonceLength != crypto_secretbox_noncebytes())
        luaL_error(L, "%s: nonce buffer should be %d bytes", kOpenName, int(crypto_secretbox_noncebytes()));
    size_t keyLength = 0;
    uint8_t* key = static_cast<uint8_t*>(luaL_checkbuffer(L, 4, &keyLength));
    if (keyLength != crypto_secretbox_keybytes())
        luaL_error(L, "%s: key buffer should be %d bytes", kOpenName, int(crypto_secretbox_keybytes()));
```
Of course we can also require 'at least' instead of 'exact'.

(side note, `crypto_secretbox_keybytes` was used in place of `crypto_secretbox_noncebytes` before)

Reply:
good catch!